### PR TITLE
Add Rotation certification Check, remove func to restart agents

### PR DESCRIPTION
**Pull request by est-suse**
#### Types of Changes ####
Test

#### Verification ####
Run the test

#### Testing ####
N/A

#### Linked Issues ####
https://github.com/rancher/qa-tasks/issues/597

#### User-Facing Change ####

```release-note
None
```

### CODE REVIEW

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 1]
```diff
@@ -313,6 +314,39 @@ var _ = Describe("Verify Create", Ordered, func() {
 				g.Expect(res).Should(ContainSubstring("local-path-test"))
 			}, "180s", "2s").Should(Succeed())
 		})
+
+		It("Verifies Certificate Rotation", func() {
+			const grepCert = "sudo ls -lt /var/lib/rancher/k3s/server/ | grep tls"
+			const expectResult = "client-ca.crt,client-ca.key,client-ca.nochain.crt,dynamic-cert.json,peer-ca.crt,peer-ca.key,server-ca.crt,server-ca.key,request-header-ca.crt,request-header-ca.key,server-ca.crt,server-ca.key,server-ca.nochain.crt,service.current.key,service.key,apiserver-loopback-client__.crt,apiserver-loopback-client__.key,"
+			var finalResult string
+			var finalerr error
```
Comment (fmoral2) [inline tests/e2e/validatecluster/validatecluster_test.go:322]: here should it be `finalErr` ?

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 2]
```diff
@@ -313,6 +314,39 @@ var _ = Describe("Verify Create", Ordered, func() {
 				g.Expect(res).Should(ContainSubstring("local-path-test"))
 			}, "180s", "2s").Should(Succeed())
 		})
+
+		It("Verifies Certificate Rotation", func() {
```
Comment (fmoral2) [inline tests/e2e/validatecluster/validatecluster_test.go:318]: @rancher-max  @est-suse   also what you guys think to separete the code in block context for be more clean? some like this : 

```
const x
const y
var finalResult string
var finalErr error

errStop := e2e.StopCluster(serverNodeNames)
Expect(errStop).NotTo(HaveOccurred(), "Server not stop correctly")

errRotate := e2e.RotateCertificate(serverNodeNames)
Expect(errRotate).NotTo(HaveOccurred(), "Certificate not rotate correctly")

errStart := e2e.StartCluster(serverNodeNames)
Expect(errStart).NotTo(HaveOccurred(), "Server not start correctly")

for _, nodeName := range serverNodeNames {
    grCert, errGrep := e2e.RunCmdOnNode(grepCert, nodeName)
    Expect(errGrep).NotTo(HaveOccurred(), "Certificate not created correctly")
...
```

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 3]
```diff
@@ -313,6 +314,39 @@ var _ = Describe("Verify Create", Ordered, func() {
...
+			const grepCert = "sudo ls -lt /var/lib/rancher/k3s/server/ | grep tls"
+			const expectResult = "client-ca.crt,client-ca.key,client-ca.nochain.crt,dynamic-cert.json,peer-ca.crt,peer-ca.key,server-ca.crt,server-ca.key,request-header-ca.crt,request-header-ca.key,server-ca.crt,server-ca.key,server-ca.nochain.crt,service.current.key,service.key,apiserver-loopback-client__.crt,apiserver-loopback-client__.key,"
```
[the remark this answers is not in the document]
Comment (fmoral2) [reply tests/e2e/validatecluster/validatecluster_test.go:320]: the formating here got out wrong, but thats the idea

**Review by dereknola — CHANGES_REQUESTED** [review]
This testing should take place in a new, seperate, E2E test. We do not want keep adding to validate_cluster.

File: tests/e2e/certrotation/certrotation_test.go [hunk 4]
```diff
@@ -0,0 +1,112 @@
...
+				"intermediate-ca.crt", "intermediate-ca.key",
+				"intermediate-ca.pem", "request-header-ca.crt", "request-header-ca.key",
+				"request-header-ca.pem", "root-ca.crt", "root-ca.key", "root-ca.pem", "server-ca.crt", "server-ca.key",
+				"server-ca.nochain.crt", "server-ca.pem", "service.current.key", "service.key", "apiserver-loopback-client__.crt",
+				"apiserver-loopback-client__.key", "",
+			}
+
+			var finalResult string
+			var finalErr error
+			errStop := e2e.StopCluster(serverNodeNames)
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:67]: same comment from here,  https://github.com/k3s-io/k3s/pull/7097#discussion_r1137461743

File: tests/e2e/certrotation/certrotation_test.go [hunk 5]
```diff
@@ -0,0 +1,112 @@
...
+			fmt.Println("OS:", *nodeOS)
+			fmt.Println("Server Nodes:", serverNodeNames)
+			fmt.Println("Agent Nodes:", agentNodeNames)
+			kubeConfigFile, err = e2e.GenKubeConfigFile(serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Verifies Certificate Rotation", func() {
+			const grepCert = "sudo ls -lt /var/lib/rancher/k3s/server/ | grep tls"
+			var expectResult = []string{"client-ca.crt", "client-ca.key", "client-ca.nochain.crt", "client-ca.pem",
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:56]: WDYTA  expectedResult ?  
and formating like this 

`	expectResult := []string{
		"client-ca.crt", "client-ca.key",
		"client-ca.nochain.crt", "client-ca.pem",
		"dynamic-cert.json", "peer-ca.crt",
		"peer-ca.key", "peer-ca.pem",
		"server-ca.crt", "server-ca.key",
		"server-ca.pem", "intermediate-ca.crt",
		"intermediate-ca.key", "intermediate-ca.pem",
		"request-header-ca.crt", "request-header-ca.key",
		"request-header-ca.pem", "root-ca.crt",
		"root-ca.key", "root-ca.pem",
		"server-ca.crt", "server-ca.key",
		"server-ca.nochain.crt", "server-ca.pem",
		"service.current.key", "service.key",
		"apiserver-loopback-client__.crt", "apiserver-loopback-client__.key",
		"",
	}`

File: tests/e2e/certrotation/certrotation_test.go [hunk 6]
```diff
@@ -0,0 +1,112 @@
...
+			if len(agentNodeNames) > 0 {
+				errRestartAgent := e2e.RestartCluster(agentNodeNames)
+				Expect(errRestartAgent).NotTo(HaveOccurred(), "Restart Agent not happened correctly")
+			}
+			Eventually(func(g Gomega) {
+				finalCert := strings.Replace(finalResult, "\n", ",", -1)
+				finalCertArray := strings.Split(finalCert, ",")
+				fmt.Println("FINAL ESTEBAN ARRAY", finalCertArray)
+				fmt.Println("FINAL ESTEBAN expectResult", expectResult)
+				Expect((finalCertArray)).Should((Equal(expectResult)), "Final certification does not match the expected results")
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:92]: here is correct Should(Equal) ? or should not ?

File: tests/e2e/certrotation/certrotation_test.go [hunk 7]
```diff
@@ -0,0 +1,112 @@
...
+	"strings"
+	"testing"
+
+	"github.com/k3s-io/k3s/tests/e2e"
+	. "github.com/onsi/ginkgo/v2"
+	. "github.com/onsi/gomega"
+)
+
+// Valid nodeOS: generic/ubuntu2004, opensuse/Leap-15.3.x86_64
+var nodeOS = flag.String("nodeOS", "generic/ubuntu2204", "VM operating system")
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:17]: here is not better merge var declarations ? 

```
var (
	// Valid nodeOS: generic/ubuntu2004, opensuse/Leap-15.3.x86_64
	nodeOS            = flag.String("nodeOS", "generic/ubuntu2004", "VM operating system")
	agentCount        = flag.Int("agentCount", 1, "number of agent nodes")
	ci                = flag.Bool("ci", false, "running on CI")
)
```

File: tests/e2e/certrotation/certrotation_test.go [hunk 8]
```diff
@@ -0,0 +1,112 @@
...
+	RunSpecs(t, "Custom Certificate Rotation Test Suite", suiteConfig, reporterConfig)
+}
+
+var (
+	kubeConfigFile  string
+	agentNodeNames  []string
+	serverNodeNames []string
+)
+
+var _ = ReportAfterEach(e2e.GenReport)
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:38]: this is not better together with functions with scope off report , after test and etc ?  ( on the EOF )

File: tests/e2e/certrotation/certrotation_test.go (hunk 6 above)
Comment (est-suse) [reply tests/e2e/certrotation/certrotation_test.go:92]: Yes!, it is correct to compare the Certificate generate against the expect result.

File: tests/e2e/certrotation/certrotation_test.go (hunk 7 above)
Comment (est-suse) [reply tests/e2e/certrotation/certrotation_test.go:17]: This is the standard use in the other vagrant files.

File: tests/e2e/certrotation/certrotation_test.go [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+
+var _ = Describe("Verify Custom CA Rotation", Ordered, func() {
+	Context("Custom CA is rotated:", func() {
+		It("Starts up with no issues", func() {
+			var err error
+			serverNodeNames, agentNodeNames, err = e2e.CreateCluster(*nodeOS, *serverCount, *agentCount)
+			Expect(err).NotTo(HaveOccurred(), e2e.GetVagrantLog(err))
+			fmt.Println("CLUSTER CONFIG")
```
[the remark this answers is not in the document]
Comment (est-suse) [reply tests/e2e/certrotation/certrotation_test.go:46]: pretty match the same, This is the standard use in the other vagrant files.

File: tests/e2e/certrotation/certrotation_test.go (hunk 5 above)
Comment (est-suse) [reply tests/e2e/certrotation/certrotation_test.go:56]: Done!

File: tests/e2e/certrotation/certrotation_test.go [hunk 10]
```diff
@@ -0,0 +1,130 @@
...
+var (
+	kubeConfigFile  string
+	agentNodeNames  []string
+	serverNodeNames []string
+)
+
+var _ = ReportAfterEach(e2e.GenReport)
+
+// RotateCertificate rotate the Certificate on each node given
+func RotateCertificate(nodeNames []string) error {
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:41]: this func shouldnt be better on a separate file for example :  "rotatecertificate.go"  ?

File: tests/e2e/certrotation/certrotation_test.go [hunk 11]
```diff
@@ -0,0 +1,130 @@
...
+				tls := re.FindAllString(grCert, -1)[0]
+				final := fmt.Sprintf("sudo diff -sr /var/lib/rancher/k3s/server/tls/ /var/lib/rancher/k3s/server/%s/ | grep -i identical | cut -f4 -d ' ' | xargs basename -a \n", tls)
+				finalResult, finalErr = e2e.RunCmdOnNode(final, nodeName)
+				Expect(finalErr).NotTo(HaveOccurred(), "Final Certification does not created correctly")
+			}
+			if len(agentNodeNames) > 0 {
+				errRestartAgent := e2e.RestartCluster(agentNodeNames)
+				Expect(errRestartAgent).NotTo(HaveOccurred(), "Restart Agent not happened correctly")
+			}
+			Eventually(func(g Gomega) {
```
Comment (fmoral2) [inline tests/e2e/certrotation/certrotation_test.go:107]: is there a need to this validation be inside a `Eventually` function ?

**Comment by dereknola** [comment]
There is not longer a need to add an additional certificate rotation E2E test. One already exists at https://github.com/k3s-io/k3s/tree/master/tests/e2e/rotateca

**Comment by dereknola** [comment]
This test is testing for **custom** certs. It calls the generate script at https://github.com/k3s-io/k3s/pull/7097/files#diff-1eab89917c6cad070cbeb0611dff6aab27f80eb2b0fed74b062c33dffd0b94c6R33-R36. This is exactly what the existing test does. If you want to also test the "regular" cert rotation, the test will need to be changed. And then I can consolidate both tests into 1 in a future PR.

**Review by fmoral2 — APPROVED** [review]

**Comment by rancher-max** [comment]
Ah yeah good call, I missed that piece. @est-suse can we update this so that it uses the built-in certs and doesn't do the custom cert generation?

File: tests/e2e/rotatecertificate/rotatecertificate_test.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+// Valid nodeOS: generic/ubuntu2004, opensuse/Leap-15.3.x86_64
+var nodeOS = flag.String("nodeOS", "generic/ubuntu2204", "VM operating system")
+var serverCount = flag.Int("serverCount", 3, "number of server nodes")
+var agentCount = flag.Int("agentCount", 1, "number of agent nodes")
+var ci = flag.Bool("ci", false, "running on CI")
+
+// Environment Variables Info:
+// E2E_RELEASE_VERSION=v1.23.1+k3s2 or nil for latest commit from master
+
+func Test_E2ECustomCARotation(t *testing.T) {
```
Comment (ShylajaDevadiga) [inline tests/e2e/rotatecertificate/rotatecertificate_test.go:25]: Lets rename this testname  as Test_E2eNormalCARotation or just Test_E2eCARotation. We dont pass user generated certs here

File: tests/e2e/rotatecertificate/vagrantfile [hunk 13]
```diff
@@ -0,0 +1,88 @@
...
+  load vagrant_defaults if File.exists?(vagrant_defaults)
+  
+  defaultOSConfigure(vm)
+  install_type = getInstallType(vm, RELEASE_VERSION, GITHUB_BRANCH)
+  
+  vm.provision "shell", inline: "ping -c 2 k3s.io"
+  
+  if role.include?("server") && role_num == 0
+    vm.provision 'file' do |scp| 
+      scp.source = '../../../contrib/util/generate-custom-ca-certs.sh'
```
Comment (ShylajaDevadiga) [inline tests/e2e/rotatecertificate/vagrantfile:30]: We don't need these for this test

File: tests/e2e/rotatecertificate/rotatecertificate_test.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+			Expect(err).NotTo(HaveOccurred(), e2e.GetVagrantLog(err))
+			fmt.Println("CLUSTER CONFIG")
+			fmt.Println("OS:", *nodeOS)
+			fmt.Println("Server Nodes:", serverNodeNames)
+			fmt.Println("Agent Nodes:", agentNodeNames)
+			kubeConfigFile, err = e2e.GenKubeConfigFile(serverNodeNames[0])
+			Expect(err).NotTo(HaveOccurred())
+		})
+
+		It("Verifies Certificate Rotation", func() {
```
Comment (ShylajaDevadiga) [inline tests/e2e/rotatecertificate/rotatecertificate_test.go:54]: Can we add this test as part of validate cluster? I suggest we don't need to create new resource for this test.

Comment (est-suse) [reply tests/e2e/rotatecertificate/rotatecertificate_test.go:54]: It has been moved.
Thanks

File: tests/e2e/rotatecertificate/rotatecertificate_test.go (hunk 12 above)
Comment (est-suse) [reply tests/e2e/rotatecertificate/rotatecertificate_test.go:25]: Delete that and use the same for validate cluster

**Review by rancher-max — CHANGES_REQUESTED** [review]
Just remove some of the extra printlns that are in here and then it LGTM

File: tests/e2e/validatecluster/validatecluster_test.go [hunk 15]
```diff
@@ -203,13 +204,16 @@ var _ = Describe("Verify Create", Ordered, func() {
 
 			Eventually(func(g Gomega) {
 				cmd := "kubectl get pods dnsutils --kubeconfig=" + kubeConfigFile
+				fmt.Println("ESTEBAN DNS ACESS", cmd)
```
Comment (ShylajaDevadiga) [inline tests/e2e/validatecluster/validatecluster_test.go:207]: Remove debug lines

**Review by dereknola — APPROVED** [review]

**Review by rancher-max — APPROVED** [review]

**Pull request merged**